### PR TITLE
[Core] Only publish WORKER_OBJECT_EVICTION when the object is out of scope or manually freed

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Currently the following sequence of events can lead to object locations being empty so pull manager fails to pull the object:

1. Driver push the generator task (5 returns) to node 1 to execute.
2. First return object is yielded and reported back to the Driver. At this point, Object `pinned_at_raylet_id` and `locations` is node 1. Node 1 also sends `SubscribeForObjectEviction`.
3. Node 1 dies.
4. PushNormalTask rpc fails and Driver retries.
5. Driver push the retried generator task to node 2 to execute.
6. First return object is yielded on node 2.
7. Driver receives `SubscribeForObjectEviction`.
8. Driver receives the node 1 death notification.
9. `ReferenceCounter::ResetObjectsOnRemovedNode` is called which triggers the `object_primary_copy_delete` callbacks.
10. `WORKER_OBJECT_EVICTION` is published to node 2 and node 2 deletes the object (This is wrong since node 2 has the new primary copy of the object).
11. Driver receives the `ReportGeneratorItemReturns` from node 2. At this point, Object `pinned_at_raylet_id` is node 2 and locations is node 2.
13.  Since in step 10, node 2 deletes the object, Driver receives the location update so now object has empty locations.

We should only publish `WORKER_OBJECT_EVICTION` when the object is out of scope or manually freed instead of when  `ReferenceCounter::ResetObjectsOnRemovedNode` is called.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rynewang — APPROVED** [review]
Is it ok for the remote to get duplicate and out of order ReportGeneratorItemReturns ?

**Comment by jjyao** [comment]
> Is it ok for the remote to get duplicate and out of order ReportGeneratorItemReturns ?

Yes.

File: src/ray/core_worker/reference_count.cc [hunk 1]
```diff
@@ -701,7 +701,7 @@ void ReferenceCounter::DeleteReferenceInternal(ReferenceTable::iterator it,
       }
     }
     // Perform the deletion.
```
Comment (alexeykudinkin) [inline src/ray/core_worker/reference_count.cc:703]: Please update comment as well

File: src/ray/core_worker/reference_count.h [hunk 2]
```diff
@@ -849,7 +849,9 @@ class ReferenceCounter : public ReferenceCounterInterface,
 
   /// Delete the object primary copy, if any. Also unsets the raylet address
   /// that the object was pinned at, if the address was set.
-  void DeleteObjectPrimaryCopy(ReferenceTable::iterator it);
+  void ResetObjectPrimaryCopy(ReferenceTable::iterator it);
+
+  void OnObjectOutOfScopeOrFreed(ReferenceTable::iterator it);
```
Comment (alexeykudinkin) [inline src/ray/core_worker/reference_count.h:854]: Do we really need this rename? The intent behind the method doesn't really change

File: src/ray/core_worker/reference_count.h [hunk 3]
```diff
@@ -849,7 +849,9 @@ class ReferenceCounter : public ReferenceCounterInterface,
 
   /// Delete the object primary copy, if any. Also unsets the raylet address
   /// that the object was pinned at, if the address was set.
-  void DeleteObjectPrimaryCopy(ReferenceTable::iterator it);
+  void ResetObjectPrimaryCopy(ReferenceTable::iterator it);
```
Comment (alexeykudinkin) [inline src/ray/core_worker/reference_count.h:852]: Need to update the comment

File: src/ray/core_worker/reference_count.h (hunk 2 above)
Comment (jjyao) [reply src/ray/core_worker/reference_count.h:854]: I think the new name make it more clear when this method should be called. The bug is that we are calling it when it shouldn't be called.

**Pull request merged**